fix: add DataSource reference handling to authorize utils

Acedus opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
A previous PR[1] introduced the concept of DataSource references as a way for a DataSource to reference another DataSource.

A DV that has its sourceRef set to a DataSource reference that resides in a different namespace would fail as authorize utils did not include the necessary changes to the clone source handler to accomodate DataSource references 

This PR fixes that and introduces a functional test for it.

[1] https://github.com/kubevirt/containerized-data-importer/pull/3760

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

arnongilboa commented on staging/src/kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1/authorize_utils.go:
```diff
@@ -45,11 +45,12 @@ func newCloneSourceHandler(dataVolume *DataVolume, dsGet dsGetFunc) (CloneSource
+		pvcSource = dataSource.Spec.Source.PVC
+		snapshotSource = dataSource.Spec.Source.Snapshot
+		if dataSource.Spec.Source.DataSource != nil {
+			pvcSource = dataSource.Status.Source.PVC
+			snapshotSource = dataSource.Status.Source.Snapshot
```
Since you are referring the Status here, can you be sure it's populated?

Acedus replied:
I can't, but the switch case after that makes it safe.

akalenyu commented:
/test pull-containerized-data-importer-e2e-nfs
The CI's host wide storage is notorious for OOMs. It dies when writing to scratch which we've never implemented O_DIRECT writing for.

Acedus commented on tests/datasource_test.go:
```diff
@@ -358,6 +367,31 @@ var _ = Describe("DataSource", func() {
 		})
 	})
+
+	Context("DataSource references", func() {
```
TBF I'm not sure how much I prefer the func test to be part of the dataimportcron section, this current func test exercises the only scenario where using DataSource references is even possible to begin with.

akalenyu reviewed (commented):
/approve

Acedus commented:
> @Acedus I apologize for being unaware of this PR and its predecessor. But my reading is that target could be namespace A, DataSource in namespace B, and source namespace C
> 
> If so, we validate that user can access namespace C but not B.
> 
> When datasource was introduced the intention was only to support jumping one namespace so:
> 
> 1. DataSource in target namespace, source in another
> 2. DataSource and source in same namespace
> 
> Do these conditions still hold true?

That's a really good point, you're right. This scenario requires more finesse, I'll adjust the PR to match that.

Acedus commented:
[Change](https://github.com/kubevirt/containerized-data-importer/compare/cd1dfa7685b777c10d79a791f980a765e8fa37c9..7a868ffce6e3997fd5d4ee3e37ed7058e57f24d6) - Addressed @mhenriks comment and added FT for it. Dropped previously added FT as logic is being tested properly as part of the new UT and FT.

Note that the previously mentioned issue of permission verification a single datasource (no intermediate) is still unresolved and will be addressed in a follow-up PR.

Acedus commented:
/hold cancel

akalenyu commented on staging/src/kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1/authorize.go:
```diff
@@ -72,6 +72,7 @@ func (dv *DataVolume) AuthorizeUser(requestNamespace, requestName string, proxy
 	if sourceNamespace == "" {
 		sourceNamespace = targetNamespace
 	}
+	intermediateDSName, intermediateDSNamespace := cloneSourceHandler.IntermediateDSName, cloneSourceHandler.IntermediateDSNamespace
```
do you want defaulting for intermediateDSNamespace in case its ""?

Acedus replied:
It's handled in the case that it's "".

Acedus replied:
We do, that's why we set it to `targetNamespace` in the case that it is `""`.
`sourceNamespace` must be a valid namespace, but not all clones require an intermediate namespace.

akalenyu commented on tests/webhook_test.go:
```diff
@@ -350,5 +369,115 @@ var _ = Describe("Clone Auth Webhook tests", func() {
+				Expect(err).ToNot(HaveOccurred())
+			},
+				Entry("when using explicit CDI permissions", explicitRole, serviceAccountName, "", false),
+				Entry("when using explicit CDI permissions without permissions on intermediate datasource", explicitRole, serviceAccountName, "", true),
```
I would pass the role that makes this succeed as part of entry args instead of the !fail bit

akalenyu reviewed (commented):
Looks good

Acedus commented:
[Change](https://github.com/kubevirt/containerized-data-importer/compare/617eae20b63dcdb253e91cf20fbcf63107ff09df..19b5bdff184a310de02cdd94a65b00eeb5d1926a) - Addressed @akalenyu's comments.

akalenyu reviewed (commented):
My approval exists on the PR so going to give mhenriks a chance to take a look too

Acedus commented:
/retest-required

Acedus commented:
/cc @mhenriks I'd appreciate your review of this.

arnongilboa commented:
@Acedus why not simply force the intermediate DataSource namespace to be the same as the pointing DataSource namespace. I see no reason to support namespace-crossing between DataSources.

Acedus commented:
> @Acedus why not simply force the intermediate DataSource namespace to be the same as the pointing DataSource namespace. I see no reason to support namespace-crossing between DataSources.

Sure, I have no objections.

Acedus commented:
fyi @akalenyu, would appreciate a re-approval.

Acedus commented:
/hold

Waiting for re-approval of @akalenyu

akalenyu commented:
/hold cancel
/approve

/test pull-containerized-data-importer-e2e-nfs
/test pull-containerized-data-importer-e2e-destructive
https://github.com/kubevirt/containerized-data-importer/pull/3809

mhenriks commented on pkg/controller/common/util.go:
```diff
@@ -393,6 +393,7 @@ var (
 
 	ErrDataSourceMaxDepthReached = errors.New("DataSource reference chain exceeds maximum depth of 1")
 	ErrDataSourceSelfReference   = errors.New("DataSource cannot self-reference")
+	ErrDataSourceCrossNamespace  = errors.New("DataSource cannot reference a DataSource in another namespace")
```
love it

mhenriks commented:
/lgtm

Acedus commented:
/cherrypick release-v1.63
